pilot/pkg/secrets/kube/multicluster.go

```diff
@@ -19,49 +19,104 @@ import (
 	"sync"
 	"time"
 
+	"istio.io/istio/pilot/pkg/features"
+
+	"go.uber.org/atomic"
+
 	"istio.io/istio/pilot/pkg/secrets"
 	"istio.io/istio/pkg/kube"
 	"istio.io/istio/pkg/kube/secretcontroller"
 	"istio.io/pkg/log"
 )
 
+type controller struct {
+	sc     *SecretsController
+	stopCh chan struct{}
+}
+
 // Multicluster structure holds the remote kube Controllers and multicluster specific attributes.
 type Multicluster struct {
-	remoteKubeControllers map[string]*SecretsController
-	m                     sync.Mutex // protects remoteKubeControllers
-	secretController      *secretcontroller.Controller
-	localCluster          string
-	stop                  chan struct{}
+	mu          sync.Mutex // protects controllers
+	controllers map[string]*controller
+
+	client            kube.Client
+	secretController  *secretcontroller.Controller
+	secretNamespace   string
+	localCluster      string
+	stop              <-chan struct{}
+	remoteSyncTimeout *atomic.Bool
 }
 
 var _ secrets.MulticlusterController = &Multicluster{}
 
-func NewMulticluster(client kube.Client, localCluster, secretNamespace string, stop chan struct{}) *Multicluster {
+func NewMulticluster(client kube.Client, localCluster, secretNamespace string) *Multicluster {
 	m := &Multicluster{
-		remoteKubeControllers: map[string]*SecretsController{},
-		localCluster:          localCluster,
-		stop:                  stop,
+		client:            client,
+		controllers:       map[string]*controller{},
+		localCluster:      localCluster,
+		secretNamespace:   secretNamespace,
+		remoteSyncTimeout: atomic.NewBool(false),
+	}
+
+	return m
+}
+
+func (m *Multicluster) addMemberCluster(clients kube.Client, key string) {
+	log.Infof("initializing Kubernetes credential reader for cluster %v", key)
+	c := &controller{
+		sc: NewSecretsController(clients, key),
+	}
+	m.mu.Lock()
+	m.controllers[key] = c
+	m.mu.Unlock()
+	if m.localCluster != key {
+		c.stopCh = make(chan struct{})
+		go clients.RunAndWait(c.stopCh)
+	} else {
+		// RunAndWait in background to avoid blocking queue
+		go clients.RunAndWait(m.stop)
 	}
+}
+
+func (m *Multicluster) Run(stop <-chan struct{}) {
+	m.stop = stop
 	// Add the local cluster
-	m.addMemberCluster(client, localCluster)
-	sc := secretcontroller.StartSecretController(client,
+	m.addMemberCluster(m.client, m.localCluster)
+	sc := secretcontroller.StartSecretController(m.client,
 		func(c kube.Client, k string) error { m.addMemberCluster(c, k); return nil },
 		func(c kube.Client, k string) error { m.updateMemberCluster(c, k); return nil },
 		func(k string) error { m.deleteMemberCluster(k); return nil },
-		secretNamespace,
+		m.secretNamespace,
 		time.Millisecond*100,
 		stop)
 	m.secretController = sc
-	return m
+
+	time.AfterFunc(features.RemoteClusterTimeout, func() {
+		m.remoteSyncTimeout.Store(true)
+	})
+
+	<-stop
```

Comment:
will this just hang until something is sent through the stop channel? why?

Reply:
https://github.com/istio/istio/blob/a776c4a0e8cda51e91ef28b2504cdf15a9fd6f92/pilot/pkg/config/kube/crdclient/client.go#L112-L124

In this instance, we block here so after `stop` is closed, we can perform cleanup. We end up closing child `stop` channels, which were given to the k8s clients. Inside client-go they will use that stop channel signal to terminate informer's watches.